Code/EnginePlugins/RecastPlugin/Components/DetourCrowdAgentComponent.cpp

```diff
@@ -0,0 +1,548 @@
+  }
+  EZ_END_PROPERTIES;
+}
+EZ_END_COMPONENT_TYPE
+// clang-format on
+
+ezDetourCrowdAgentComponent::ezDetourCrowdAgentComponent()
+{
+  m_uiTargetDirtyBit = 0;
+  m_uiSteeringFailedBit = 0;
+  m_uiErrorBit = 0;
+  m_uiParamsDirtyBit = 0;
+}
+
+ezDetourCrowdAgentComponent::~ezDetourCrowdAgentComponent() = default;
+
+void ezDetourCrowdAgentComponent::SerializeComponent(ezWorldWriter& stream) const
+{
+  SUPER::SerializeComponent(stream);
+  ezStreamWriter& s = stream.GetStream();
+
+  s << m_fRadius;
+  s << m_fHeight;
+  s << m_fMaxSpeed;
+  s << m_fMaxAcceleration;
+  s << m_fStoppingDistance;
+  s << m_MaxAngularSpeed;
+  s << m_RotationMode;
+  s << m_fPushiness;
+}
+
+void ezDetourCrowdAgentComponent::DeserializeComponent(ezWorldReader& stream)
+{
+  SUPER::DeserializeComponent(stream);
+  const ezUInt32 uiVersion = stream.GetComponentTypeVersion(GetStaticRTTI());
+  ezStreamReader& s = stream.GetStream();
+
+  s >> m_fRadius;
+  s >> m_fHeight;
+  s >> m_fMaxSpeed;
+  s >> m_fMaxAcceleration;
+  s >> m_fStoppingDistance;
+  s >> m_MaxAngularSpeed;
+  s >> m_RotationMode;
+  s >> m_fPushiness;
+}
+
+void ezDetourCrowdAgentComponent::FillAgentParams(ezDetourCrowdAgentParams& out_params) const
+{
+  out_params.m_fRadius = m_fRadius;
+  out_params.m_fHeight = m_fHeight;
+  out_params.m_fMaxSpeed = m_fMaxSpeed;
+  out_params.m_fMaxAcceleration = m_fMaxAcceleration;
+  out_params.m_fSeparationWeight = m_fPushiness;
+}
+
+void ezDetourCrowdAgentComponent::SetRadius(float fRadius)
+{
+  if (fRadius < 0)
+    fRadius = 0;
+
+  if (fRadius == m_fRadius)
+    return;
+
+  m_fRadius = fRadius;
+  m_uiParamsDirtyBit = 1;
+}
+
+void ezDetourCrowdAgentComponent::SetHeight(float fHeight)
+{
+  if (fHeight < 0.01f)
+    fHeight = 0.01f;
+
+  if (fHeight == m_fHeight)
+    return;
+
+  m_fHeight = fHeight;
+  m_uiParamsDirtyBit = 1;
+}
+
+void ezDetourCrowdAgentComponent::SetMaxSpeed(float fMaxSpeed)
+{
+  if (fMaxSpeed < 0.0f)
+    fMaxSpeed = 0.0f;
+
+  if (fMaxSpeed == m_fMaxSpeed)
+    return;
+
+  m_fMaxSpeed = fMaxSpeed;
+  m_uiParamsDirtyBit = 1;
+}
+
+void ezDetourCrowdAgentComponent::SetMaxAcceleration(float fMaxAcceleration)
+{
+  if (fMaxAcceleration < 0.0f)
+    fMaxAcceleration = 0.0f;
+
+  if (m_fMaxAcceleration == fMaxAcceleration)
+    return;
+
+  m_fMaxAcceleration = fMaxAcceleration;
+  m_uiParamsDirtyBit = 1;
+}
+
+void ezDetourCrowdAgentComponent::SetStoppingDistance(float fStoppingDistance)
+{
+  if (fStoppingDistance < 0.001f)
+    fStoppingDistance = 0.001f;
+
+  m_fStoppingDistance = fStoppingDistance;
+}
+
+void ezDetourCrowdAgentComponent::SetMaxAngularSpeed(ezAngle maxAngularSpeed)
+{
+  if (maxAngularSpeed.GetRadian() < 0.0f)
+    maxAngularSpeed.SetRadian(0.0f);
+
+  m_MaxAngularSpeed = maxAngularSpeed;
+}
+
+void ezDetourCrowdAgentComponent::SetPushiness(float fPushiness)
+{
+  if (fPushiness < 0.0f)
+    fPushiness = 0.0f;
+
+  if (fPushiness != m_fPushiness)
```

Comment:
==

Reply:
thanks! i should stop writing code at nighttime